data/AlexTemplate.hs

```diff
@@ -161,11 +161,15 @@ alex_scan_tkn user__ orig_input len input__ s last_acc =
         let
                 base   = alexIndexInt32OffAddr alex_base s
                 offset = PLUS(base,ord_c)
-                check  = alexIndexInt16OffAddr alex_check offset
 
-                new_s = if GTE(offset,ILIT(0)) && EQ(check,ord_c)
-                          then alexIndexInt16OffAddr alex_table offset
-                          else alexIndexInt16OffAddr alex_deflt s
+                new_s
+                  | GTE(offset,ILIT(0))
+                  , check <- alexIndexInt16OffAddr alex_check offset
+                  , EQ(check,ord_c)
+                  = alexIndexInt16OffAddr alex_table offset
+
+                  | otherwise
+                  = alexIndexInt16OffAddr alex_deflt s
```

Comment:
Maybe it is too early in the morning and my brain hasn't woken up... but I cannot see the difference between the new and the old code.

Reply:
I should have explained more ;)

The issue is that 'check' is unlifted, hence it's a strict binding. The way it was written before, it was evaluated before checking that `offset` was greater than or equal to 0. I've just moved the binding after the check.

Adding a test case for this seems difficult. We could add an assertion in the indexing function in debug mode that the offset is positive, but that seemed overkill. I've found this bug in the lexer code generated for Cabal-syntax. Any use of this code with the JS backend fails (the JS backend doesn't have an addressable heap, instead 'alex_check' array is represented as a JS array, so out-of-bound access triggers an exception).

Reply:
Ok that is subtle indeed.  I unlearned ML a long time ago... ;-)

- Could you add your explanation to the commit message?
- Could you add a CHANGELOG entry for this?
- Do you need a release of the fix?

Reply:
Done.

A release would be nice as I'll have to fix Cabal-syntax's lexer.